### PR TITLE
Puzzle contract

### PR DESCRIPTION
**Pull request by alajko**
## Issue

Puzzle Smart Contract
 
Will refactor restclient according to new contract, if contract looks ok (i.e refactor https://github.com/harmony-one/harmony/pull/793/files)

### CODE REVIEW

File: contracts/Puzzle.sol [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+    constructor() public {
+        manager = msg.sender;
+    }
+
+    /**
+     * @dev The player enters into the current game session by
+     *      paying at least 2 token.
+     */
+    function play() public payable {
+        require(msg.value == 2 ether, INSUFFICIENT_FUND_MESSAGE);
```
Comment (rlan35) [inline contracts/Puzzle.sol]: >=?

File: contracts/Puzzle.sol [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+        players.push(msg.sender);
+    }
+
+    /**
+     * @dev returns the current best level the player has crossed   
+     */
+    function getlevel() public view returns(uint) {
```
Comment (rlan35) [inline contracts/Puzzle.sol]: getLevel

File: contracts/Puzzle.sol [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+        return bestlevel[msg.sender];
+    }
+
+    /**
+     * @dev returns the current best level the player has crossed   
+     */
+    function setlevel(address player, uint level) public restricted{
```
Comment (rlan35) [inline contracts/Puzzle.sol]: setLevel

File: contracts/Puzzle.sol [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+       bestlevel[player] = level;
+    }
+
+    /**
+     * @dev pay the player if they have crossed their last best level. Set tjeir
+     */
+    function payout(address payable player, uint level) public payable restricted {
+        if ( level > bestlevel[player]) {
```
Comment (rlan35) [inline contracts/Puzzle.sol]: extra space before "level"

File: contracts/Puzzle.sol [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+        players.length = 0;
+    }
+    
+    /**
+     * @dev Returns a list of all players in the current session.
+     */
+    function getPlayers() public view returns (address payable[] memory) {
+        return players;
+    }
+}
```
Comment (rlan35) [inline contracts/Puzzle.sol]: add new line.

File: contracts/Puzzle.sol [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+    function setlevel(address player, uint level) public restricted{
+       bestlevel[player] = level;
+    }
+
+    /**
+     * @dev pay the player if they have crossed their last best level. Set tjeir
+     */
+    function payout(address payable player, uint level) public payable restricted {
+        if ( level > bestlevel[player]) {
+            amount = level - bestlevel[player];
```
Comment (LeoHChen) [inline contracts/Puzzle.sol]: amount is way too less.  should be calculated dynamically.  Please use a function to calculate the amount.

File: contracts/Puzzle.sol [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+    modifier restricted() {
+        require(msg.sender == manager, RESTRICTED_MESSAGE);
+        _;
+    }
+    
+    /**
+     * @dev Resets the current session by deleting the players
+     */
+    function reset() public restricted {
+        for (uint i=0; i < players.length; i++){
```
Comment (LeoHChen) [inline contracts/Puzzle.sol]: is it a good way of clean all player data?

File: contracts/Puzzle.sol [hunk 8]
```diff
@@ -0,0 +1,69 @@
...
+
+    modifier restricted() {
+        require(msg.sender == manager, RESTRICTED_MESSAGE);
+        _;
+    }
+    
+    /**
+     * @dev Resets the current session by deleting the players
+     */
+    function reset() public restricted {
```
Comment (LeoHChen) [inline contracts/Puzzle.sol]: we should also provide a way for user to reset the level, by paying like 1 token.

File: contracts/Puzzle.sol (hunk 1 above)
Comment (alajko) [reply contracts/Puzzle.sol]: I actually did not want user to pay more than 2!

File: contracts/Puzzle.sol (hunk 7 above)
Comment (alajko) [reply contracts/Puzzle.sol]: Yes, standard way afaict in solidity.

File: contracts/Puzzle.sol (hunk 5 above)
Comment (alajko) [reply contracts/Puzzle.sol]: done

File: contracts/Puzzle.sol (hunk 4 above)
Comment (alajko) [reply contracts/Puzzle.sol]: done

File: contracts/Puzzle.sol (hunk 3 above)
Comment (alajko) [reply contracts/Puzzle.sol]: done -- weird this was left like this.

File: contracts/Puzzle.sol (hunk 8 above)
Comment (alajko) [reply contracts/Puzzle.sol]: this is for session reset.

Comment (alajko) [reply contracts/Puzzle.sol]: but reset for level is free!

File: contracts/Puzzle.sol (hunk 6 above)
Comment (alajko) [reply contracts/Puzzle.sol]: okay!

File: contracts/Puzzle.sol (hunk 1 above)
Comment (alajko) [reply contracts/Puzzle.sol]: will change it.

File: contracts/Puzzle.sol [hunk 9]
```diff
@@ -0,0 +1,98 @@
...
+     *      paying at least 2 token.
+     */
+    function play() public payable returns (uint) {
+        require(msg.value >= 2 ether, INSUFFICIENT_FUND_MESSAGE);
+        sessionID = random();
+        playerLevel[msg.sender] = 0;
+        playerSession[msg.sender] = sessionID;
+        playerStake[msg.sender] = msg.value;
+        players.push(msg.sender);
+        return sessionID;
```
Comment (LeoHChen) [inline contracts/Puzzle.sol]: is sessionID the txid?

File: contracts/Puzzle.sol [hunk 10]
```diff
@@ -0,0 +1,98 @@
...
+    /**
+     * @dev pay the player if they have crossed their last best level.
+     */
+    function payout(address payable player, uint level, uint session, string memory /*sequence*/) public payable restricted {
+            require(playerSession[player] == session, INCORRECT_SESSION );
+            require(level > playerLevel[player], INCORRECT_LEVEL);
+            progress = level - playerLevel[player]; //if a later transaction for a higher level comes earlier.
+            setLevel(player,level);
+            amount = progress*(playerStake[player]/thresholdLevel);
+            player.transfer(amount);
```
Comment (LeoHChen) [inline contracts/Puzzle.sol]: will you have sufficient amount of money for the transfer? where is it coming from?

Comment (alajko) [reply contracts/Puzzle.sol]: Need to fund it like we fund faucet contract.

File: contracts/Puzzle.sol (hunk 9 above)
Comment (alajko) [reply contracts/Puzzle.sol]: No, its just a randomness we generate to maintain a session.

File: contracts/Puzzle.sol (hunk 8 above)
Comment (rlan35) [reply contracts/Puzzle.sol]: Why do we need to delete all users at once?

**Pull request merged**